[Feat/게시글작성] 게시글 작성

IToriginal opened the pull request:
## 🤔 Motivation
- 게시글 작성 API 코드 개발 완료

<br>

## 💡 Key Changes
- [🔨 fix: 불필요한 코드 삭제](https://github.com/four-uncles/saramara-community-server/commit/d80e8250953400bcd49939114788cca2e7bb4b3a)
 📍 저번 PR에서 놓쳤던 현재 개발 시점에서 불필요한 Page 처리에 대한 코드를 추가적으로 발견하여 삭제하였습니다.
- [✨](https://github.com/four-uncles/saramara-community-server/commit/935747616f8aae83885675589b82b3eb6ccb1666) [feat(#26): 게시글 작성 API](https://github.com/four-uncles/saramara-community-server/commit/935747616f8aae83885675589b82b3eb6ccb1666)
 📍 게시글 작성 API에 필요한 RequestDto, Controller, Service 코드를 구현하였습니다.

### API TEST for PostMan: 포스트맨 프로그램을 통해 테스트를 진행하였습니다.
아래의 이미지는 테스트에 사용된 JSON 값과 요청에 대한 API 응답 값입니다.
응답 처리는 API 명세서에 명시된 것처럼 "msg", "code", "data" 순으로 처리 하였습니다.

<img width="1012" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/0fd3214f-cb17-4f78-9466-1a36aee15608">

<img width="1013" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/a9410e70-5760-429a-a2a6-b4639b1694d6">

------
- 🔨 Code Review 반영(94ce0702f06dc8a0b0f6eb5395f7d065ccf3e987): Category -> CategoryBoard 변경
<img width="1485" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/cce7b586-d838-44ae-81fb-a2f834cfef21">

- 🐞 Bugfix: deadline 요청이 제대로 동작하지 않는 Bug 발견(https://github.com/four-uncles/saramara-community-server/pull/30#issuecomment-1607178625)


<br>

## 👨‍👨‍👧‍👦 To Reviewers
- @Jooney-95 @byeongJoo05 @hb9397 

close: #26

IToriginal committed:
:hammer: fix: 불필요한 코드 삭제

IToriginal committed:
:sparkles: feat(#26): 게시글 작성 API

IToriginal committed:
:wrench:  fix(#26): JWT 검증 path 수정

- 원할한 테스트 진행을 위해 SecurityConfig에 임시로 path 지정
- board, comment의 경우 회원의 개발이 아직 미완성 단계이므로 회원의 로직이 구현되면 임시로 지정한 path를 삭제하고 JWT 검증을 함께 테스트할 예정

IToriginal committed:
:hammer: fix(#26): Dto Colume 값 수정 및 좋아요 수 기본값 지정

- 게시글 작성시 Dto에 필요한 값 수정 (memberId를 Join해서 오기 때문에 이메일과 닉네임을 Dto로 둘 이유가 없다고 판단하였음)
- @PerPersist에 좋아요 수에 대한 기본값 0 지정 코드가 누락되어 코드 추가

byeongJoo05 commented on src/main/java/com/kakao/saramaracommunity/board/controller/BoardController.java:
```diff
@@ -0,0 +1,40 @@
+    private final BoardService boardService;
+
+    @PostMapping("/register")
+    public ResponseEntity<?> createBoard(@RequestBody @Valid BoardRequestDto.SaveRequestDto requestDto) {
```
혹시 응답 데이터에 대한 자료형 변경이 없다면 `wildcard`가 아닌 직접 명시해주는건 어떨까요?

컴파일 시간에 대해 큰 차이가 벌어져보이진 않지만, 타입체크가 일어날 시간이 생길 거 같아서요!

IToriginal replied:
해당 부분을 잘 모르고 있어서 어떻게 해야할지 어려움을 느끼네요.🥲 이 부분에 대해 아시는 바가 있으시면 공유 부탁드릴게요! 저도 이 부분에 대해 조금 더 학습을 한 이후에 Meeting 날짜나 별도의 시간에 생각을 공유하고 코드를 수정하겠습니다!

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/controller/BoardController.java:
```diff
@@ -0,0 +1,40 @@
+    private final BoardService boardService;
+
+    @PostMapping("/register")
+    public ResponseEntity<?> createBoard(@RequestBody @Valid BoardRequestDto.SaveRequestDto requestDto) {
```
createBoard 메서드에서 요청 파라미터에 `BoardRequestDto.SaveRequestDto requestDto` 이 구문을 따라 확인해보니
BoardRequestDto 클래스의 생성자로 SaveRequestDto가 만들어진 것을 확인했습니다. 이렇게 요청 파라미터 구조를 개발하신 이유가 궁금합니다!!

byeongJoo05 replied:
저도 읽어봐서 궁금한건데 `static 내부 클래스` 로 dto 나눠놓으신거 같은데 혹시 추후 확장 개념으로 사용하신건가요?

그러면 공통으로 사용될 변수를 외부 클래스로 빼놓는게 나을듯 싶습니다!

IToriginal replied:
확장성을 고려했다기 보다는 파일을 정리하기 위해서 클래스 내부에 내부클래스(inner Class)로 관리할 수 있도록 작성했습니다.
처음에는 클래스로 구분해서 나눌 생각이였지만, 오히려 class File이 많아져서 복잡해질 수 있다 판단하였고, 저희 프로젝트는 CRUD에 대한 요청만 있을 것으로 판단해서 1개의 class File에 내부 클래스로 관리를 하고, 이러한 방법이 선택적인 방법이지만 제가 찾아본 내용은 내부 클래스로 정의하는 것이 코드의 구조화와 캡슐화를 위해 사용 되고, 아래와 같은 내부 클래스의 이점사항이 있어서 이러한 방식을 이용해서 개발하게 되었습니다.

> #### 👨🏻‍💻 내부 클래스를 사용하는 이점 
> 1. 코드 구조화: 내부 클래스를 사용하면 연관된 클래스를 논리적으로 그룹화할 수 있다. SaveRequestDto는 BoardRequestDto에 관련된 기능을 제공하므로, 클래스 내에서 함께 선언함으로써 관련된 코드를 한 곳에 위치시킬 수 있다.
>
> 2. 접근 제어: 내부 클래스를 사용하면 외부 클래스에서 내부 클래스의 멤버에 직접 접근하는 것을 제한할 수 있다. 내부 클래스를 private으로 선언하거나, 필요에 따라 public, protected 등의 접근 제어자를 사용하여 외부에서의 접근을 제한할 수 있다.
>
> 3. 코드 응집성: 내부 클래스를 사용하면 클래스 간의 관련된 데이터와 동작을 한 곳에 유지할 수 있다. SaveRequestDto가 BoardRequestDto의 일부로 정의되어 있기 때문에, SaveRequestDto의 데이터와 동작은 BoardRequestDto와 밀접하게 연관되어 있을 것이다.

IToriginal replied:
@byeongJoo05 오히려 반대의 생각으로 사용했습니다.
dto을 사용하는데 class 파일이 많으면 오히려 더 복잡해질 가능성을 생각했습니다.
그래서 기본적인 CRUD의 Request나 Respoonse의 경우, 제가 사용한 내부 클래스의 방법으로 하는 것이 하나의 파일로 관리하고 한 class에 코드가 묶여있기 때문에 오히려 유지보수에 용이할 것이라 판단했습니다.
제가 생각하기에는 아직은 생각하지 못한 CRUD를 제외한 다른 목적을 가진 Dto가 생긴다면 그 목적에 맞는 class를 하나 생성하는게 관리하기 좋다고 생각합니다

langoustinee replied:
@IToriginal 저번 미팅 때 나왔던 이야기로는 일단 요청이나 응답 관련 DTO는 클래스별로 일일이 만들어서 개발하고 이후에 만들어진 DTO들을 관리하는 일정을 잡기로 했던 것으로 기억하고 있어서 여쭤봤습니다!

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/controller/BoardController.java:
```diff
@@ -0,0 +1,40 @@
+        Board savedBoard = boardService.saveBoard(requestDto);
+
+        // 응답 데이터 생성
+        Map<String, Object> response = new HashMap<>();
+        response.put("code", 200);
+        response.put("msg", "success");
+
+        // 게시글 정보
+        response.put("data", savedBoard);
+
+        return ResponseEntity.ok(response);
```
혹시 시간이 남으시다면 응답형식을 위처럼 Map 객체를 생성해서 전달하기 보단 ResponseEntity의 body 메서드를 이용하는 건 어떨까요?

IToriginal replied:
Common으로 공통으로 사용할 method를 생성해서 동일한 패턴의 응답을 반환하도록 코드를 수정해보겠습니다. 🤔

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/dto/request/BoardRequestDto.java:
```diff
@@ -0,0 +1,36 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+public class BoardRequestDto {
+
+    @AllArgsConstructor
+    @NoArgsConstructor
+    @Getter
+    @Builder
+    public static class SaveRequestDto {
+
+        @NotBlank(message = "제목을 입력해주세요.")
+        private String title;
+
+        @NotBlank(message = "내용을 입력해주세요.")
+        private String content;
+
+        @NotNull(message = "카테고리를 선택해주세요.")
+        private Category category;
+
+        @NotNull(message = "사용자의 고유번호가 존재하지 않습니다.")
+        private Long memberId;
+
+        private LocalDateTime deadLine;
+    }
+}
```
요기 클래스 구조가 생소한 부분이 있어서 개발 의도를 설명받고 싶습니다! 시간나시면 공유 부탁드립니다.

IToriginal replied:
내부 클래스를 사용한 의도가 궁금하신건가요? https://github.com/four-uncles/saramara-community-server/pull/30#discussion_r1241907456 위의 리뷰와 비슷한 맥락인거 같아서 위의 댓글로 궁금점이 해소될 수 있을까요? 🤔

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/entity/Category.java:
```diff
@@ -1,5 +1,5 @@
 package com.kakao.saramaracommunity.board.entity;
 
-public enum CategoryBoard {
+public enum Category {
```
Enum은 웬만하면 이전 이름인 CategoryBoard를 고수하는게 좋을 것 같습니다!
제가 정답은 아닌데 보통 도메인과 타입을 같이 결합하여 명명하는 것이 좋다고 알고 있습니다.

IToriginal replied:
이 부분에 대해 네이밍을 어떻게 하면 좋을까 고민이 많았는데 이전의 네이밍을 명확하게 하는게 아무래도 좋겠네요. 감사합니다. 🙏

IToriginal replied:
반영 완료했습니다! 확인 부탁합니다

IToriginal committed:
:bug: Bugfix(#26): deadLine 요청이 동작하지 않는 버그 수정

- 지정된 유효기간의 날짜와 시간이 데이터에 들어가지 않는 버그 현상을 발견하고 해당 로직을 수정
- (추가)Category의 비효율적인 처리 코드를 발견하고 이를 수정

IToriginal committed:
:hammer: fix(#26): Category 네이밍 수정 - Code Review 반영

IToriginal commented:
### Bugfix: deadLine의 요청 값이 제대로 작동하지 않는 부분을 확인
- 지정한 유효기간을 작성해서 요청을 보냈음에도 불구하고 null이 저장되는 Bug를 발견해서 이를 수정하였습니다.
> <img width="480" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/1bad48a2-2bca-4c68-b45b-0d127f099191">

- 수정을 완료하고 다시 포스트맨을 이용해서 테스트를 확인한 결과 제대로 동작하는 것을 확인했습니다.
> <img width="391" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/13b92c09-5b0a-45d7-9c2d-57fbec4e6ff0">

langoustinee commented:
제가 댓글로 질문하는 사항들에 대해서는 저도 정답이라고 확신을 가지고 질문하는 것이 아니라 크게 영향받지 않으셨음 합니다~
